presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CrossJoinWithArrayNotContainsToAntiJoin.java

```diff
@@ -0,0 +1,213 @@
+
+    Metadata metadata;
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public CrossJoinWithArrayNotContainsToAntiJoin(Metadata metadata, FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
+
+    public static RowExpression getCandidateArrayNotContainsExpression(FunctionResolution functionResolution, RowExpression filterPredicate, List<VariableReferenceExpression> leftInput, List<VariableReferenceExpression> rightInput)
+    {
+        List<RowExpression> conjuncts = extractConjuncts(filterPredicate);
+        for (RowExpression conjunct : conjuncts) {
+            if (PlannerUtils.isNegationExpression(functionResolution, conjunct) &&
+                    isSupportedArrayContainsFilter(functionResolution, conjunct.getChildren().get(0), leftInput, rightInput)) {
+                return conjunct;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isRewriteCrossJoinArrayNotContainsToAntiJoinEnabled(session);
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        JoinNode joinNode = captures.get(JOIN);
+
+        if (!(joinNode.getType().equals(JoinNode.Type.INNER) && joinNode.getCriteria().isEmpty())) {
+            return Result.empty();
+        }
+        List<VariableReferenceExpression> leftColumns = joinNode.getLeft().getOutputVariables();
+        List<VariableReferenceExpression> rightColumns = joinNode.getRight().getOutputVariables();
+        RowExpression filterExpression = node.getPredicate();
+        FunctionResolution functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+
+        RowExpression arrayNotContainsExpression = getCandidateArrayNotContainsExpression(functionResolution, filterExpression, leftColumns, rightColumns);
+        if (arrayNotContainsExpression == null) {
+            return Result.empty();
+        }
+        List<RowExpression> allConjuncts = extractConjuncts(filterExpression);
+        List<RowExpression> remainingConjuncts = allConjuncts.stream().filter(x -> !x.equals(arrayNotContainsExpression)).collect(Collectors.toList());
+
+        RowExpression arrayContainsExpression = arrayNotContainsExpression.getChildren().get(0);
+        RowExpression array = ((CallExpression) arrayContainsExpression).getArguments().get(0);
+        RowExpression element = ((CallExpression) arrayContainsExpression).getArguments().get(1);
+        boolean arrayAtLeftInput = leftColumns.containsAll(VariablesExtractor.extractAll(array));
```

Comment:
Similar, I think `array` and `element` are variable reference now, should we change here accordingly?

Reply:
fixed and added asserts